### DISCUSSION TITLE
Login a user manually with the new authentication system

### DISCUSSION THREAD

**Discussion by bernard-ng** [symfony/symfony#42397] [Q&A] [answered]
How do I manually login a user with the new authentication system ?
I'm used to

```
$guardHandler->authenticateUserAndHandleSuccess($user, $request, $authenticator, 'main'); 
```

But this is deprecated since Symfony 5.3 and I can't find an alternative way in the doc
any idea ? Thanks

**Comment by apfelbox** [comment]
Here is some discussion about that and currently working code: https://github.com/symfony/symfony/issues/40662

**Comment by welcoMattic** [comment, marked as answer]
You can inject `Symfony\Component\Security\Http\Authentication\AuthenticatorManagerInterface` in your controller and use `$authenticatorManager->authenticateUser($user, $authenticator, $request)` to authenticate your user.

With this, it's up to you to retrieve the user.

 The `$authenticator` is the authenticator that you want to use to authenticate your user (it could be `Symfony\Component\Security\Http\Authenticator\FormLoginAuthenticator` for example.

**Reply by bernard-ng** [reply]
Thanks

**Reply by Saaaam** [reply]
Apparently, `Symfony\Component\Security\Http\Authentication\AuthenticatorManagerInterface` is depreciated.

We shall use  `Symfony\Component\Security\Http\Authentication\UserAuthenticatorInterface` instead...

**Reply by lamkimphu258** [reply]
Hi @Saaaam, why do you know that `Symfony\Component\Security\Http\Authentication\AuthenticatorManagerInterface` is deprecated? I don't see they talk about that in code.

**Reply by ahundiak** [reply]
It is actually the older Guard\AuthenticatorInterface that is deprecated.  The Http stuff is the newer stuff.  Hopefully it will last at least a year or two before being replaced.

**Reply by Saaaam** [reply]
> 
> 
> Hi @Saaaam, why do you know that `Symfony\Component\Security\Http\Authentication\AuthenticatorManagerInterface` is deprecated? I don't see they talk about that in code.

That is written when I check the Service with `php bin/console debug:autowiring`

**Reply by lamkimphu258** [reply]
@Saaaam when I run `php bin/console debug:autowiring`, I just see this  `Symfony\Component\Security\Core\Authentication\AuthenticationManagerInterface (security.authentication.manager) - deprecated
`. Do you confuse or I miss anything?

**Reply by lamkimphu258** [reply]
@Saaaam thank you :smile:

**Comment by arturdoruch** [comment]
This is my solution, truly manual logging.

```php
<?php

namespace App\Security;

use Symfony\Bundle\SecurityBundle\Security\FirewallMap;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\RequestStack;
use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
use Symfony\Component\Security\Core\User\UserInterface;

class ManualAuthenticator
{
    private $firewallMap;
    private $tokenStorage;
    private $session;

    public function __construct(FirewallMap $firewallMap, TokenStorageInterface $tokenStorage, RequestStack $requestStack)
    {
        $this->firewallMap = $firewallMap;
        $this->tokenStorage = $tokenStorage;
        $this->session = $requestStack->getSession();
    }

    /**
     * Manually login the user.
     *
     * @param UserInterface $user
     * @param Request $request
     */
    public function authenticate(UserInterface $user, Request $request)
    {
        $firewallName = $this->firewallMap->getFirewallConfig($request)->getName();
        $token = new UsernamePasswordToken($user, $user->getPassword(), $firewallName, $user->getRoles());
        $this->tokenStorage->setToken($token);
        $this->session->set('_security_' . $firewallName, serialize($token));
    }
}
```

In controller action

```php

use App\Security\ManualAuthenticator;

public function action(Request $request, ManualAuthenticator $manualAuthenticator)
{
    $user = '';
    $manualAuthenticator->authenticate($user, $request);

    // Redirect to some page after login.
    //return $this->redirectToRoute();
}
```

**Comment by Saaaam** [comment]
Yes, this is it ;) Look at the end of the line

--
Damien Poncin

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
